python-stdlib/datetime/datetime.py

```diff
@@ -0,0 +1,449 @@
+
+    def __floordiv__(self, other):
+        if isinstance(other, timedelta):
+            return self._s // other._s
+        else:
+            return timedelta(seconds=int(self._s // other))
+
+    def __mod__(self, other):
+        return timedelta(seconds=self._s % other._s)
+
+    def __divmod__(self, other):
+        q, r = divmod(self._s, other._s)
+        return q, timedelta(seconds=r)
+
+    def __neg__(self):
+        return timedelta(seconds=-self._s)
+
+    def __pos__(self):
+        return self
+
+    def __eq__(self, other):
+        return self._s == other._s
+
+    def __le__(self, other):
+        return self._s <= other._s
+
+    def __lt__(self, other):
+        return self._s < other._s
+
+    def __ge__(self, other):
+        return self._s >= other._s
+
+    def __gt__(self, other):
+        return self._s > other._s
+
+    def __bool__(self):
+        return self._s != 0
+
+    def __neg__(self):
+        return timedelta(seconds=-self._s)
+
+    def __pos__(self):
+        return self
+
+    def __abs__(self):
+        return timedelta(seconds=abs(self._s))
+
+    def __repr__(self):
+        return "datetime.timedelta(seconds=%d)" % self._s
+
+    def __str__(self):
+        return self.isoformat()
+
+    def isoformat(self):
+        t = self.tuple()
+        if 0 <= self._s < 86400:
+            return "%02d:%02d:%02d" % t[2:]
+        else:
+            return "%s%dd %02d:%02d:%02d" % t
+
+    def tuple(self, sign_pos=""):
+        s = self._s
+        if s < 0:
+            s *= -1
+            g = "-"
+        else:
+            g = sign_pos
+        m, s = divmod(s, 60)
+        h, m = divmod(m, 60)
+        d, h = divmod(h, 24)
+        return g, d, h, m, s
+
+
+timedelta.min = timedelta(seconds=-2147483648)  # -2**31
+timedelta.max = timedelta(seconds=2147483647)  #  2**31 - 1
+timedelta.resolution = timedelta(seconds=1)
+
+
+class timezone:
+    def __init__(self, offset, name=None):
+        if not (timedelta(hours=-24) < offset < timedelta(hours=24)):
+            raise ValueError
+        self._offset = offset
+        self._name = name
+
+    def __str__(self):
+        return self.tzname(None)
+
+    def utcoffset(self, dt):
+        dst = self.dst(dt)
+        if dst is None or not dst:
+            return self._offset
+        return self._offset + dst
+
+    def dst(self, dt):
+        return None
+
+    def tzname(self, dt):
+        if self._name != None:
+            return self._name
+        return self.isoformat(dt)
+
+    def isoformat(self, dt, *, utc=True):
+        td = self.utcoffset(dt)
+        if utc and not td:
+            return "UTC"
+        sign, day, hour, minute, second = td.tuple("+")
+        return "%s%s%02d:%02d" % ("UTC" if utc else "", sign, hour, minute)
+
+
+timezone.utc = timezone(timedelta(0))
+
+
+class datetime:
+    MINYEAR = 1
+    MAXYEAR = 9999
+
+    def __init__(self, year, month, day, hour=0, minute=0, second=0, tzinfo=None):
```

Comment:
`timedelta` resolution is now 1 nanosecond, instead of 1 second. The module is now suitable for logging at microsecond scale and below. As a bonus, `timedelta` can now handle deltas which span ±292 years (instead of the previous ±68 years). The cost is 4 more bytes per instance.

I chose a single 64-bit `int`, holding the number of nanoseconds, instead of two 32-bit `int`s, holding the number of second and the number of micro- or nanoseconds, because the implementation was straightforward. I chose nanoseconds instead of microseconds because with it makes more sense to me working with a tiny resolution instead of big time ranges on microcontrollers. It's easy to switch to microseconds (and ±292471 years) if you like.

Reply:
> `timedelta` resolution is now 1 nanosecond, instead of 1 second.
> ...
> I chose a single 64-bit `int`, holding the number of nanoseconds,

Looking at the code, it looks like it uses a single "big integer" stored in `._ns`.  This will allow arbitrarily large values so there shouldn't be any limit on the time span of the delta.

> I chose nanoseconds instead of microseconds because with it makes more sense to me working with a tiny resolution instead of big time ranges on microcontrollers. It's easy to switch to microseconds (and ±292471 years) if you like.

I think it's best to keep it at microsecond resolution, to match CPython.  That will also keep the `._ns` values smaller and so use a bit less memory.

For really accurate timing (sub-microsecond resolution) it's probably better to use `time.time_ns()`, `time.monotonic_ns()` and subtraction of these values to compute time differences.

Reply:
OK. I converted the code from nanoseconds to microseconds.